Add doc for analyzers CA1839 and CA1840 (#27249)
dotnet/docs: docs/fundamentals/code-analysis/quality-rules/ca1840.md

---
title: "CA1840: Use 'Environment.CurrentManagedThreadId'"
description: "'Environment.CurrentManagedThreadId' is simpler and faster than 'Thread.CurrentThread.ManagedThreadId'."
ms.date: 11/23/2021
ms.topic: reference
f1_keywords:
  - "CA1840"
helpviewer_keywords:
  - "CA1840"
author: buyaa-n
dev_langs:
- CSharp
- VB
---

# CA1840: Use `Environment.CurrentManagedThreadId` instead of `Thread.CurrentThread.ManagedThreadId`

| | Value |
|-|-|
| **Rule ID** |CA1840|
| **Category** |[Performance](performance-warnings.md)|
| **Fix is breaking or non-breaking** |Non-breaking|

## Cause

> Using `Thread.CurrentThread.ManagedThreadId` for getting the current managed thread id instead of `Environment.CurrentManagedThreadId`.

Reviewer: ```suggestion
Using `Thread.CurrentThread.ManagedThreadId` for getting the current managed thread ID instead of <xref:System.Environment.CurrentManagedThreadId?displayProperty=fullName>.
```

## Rule description

> `Environment.CurrentManagedThreadId` is more compact and efficient replacement of `Thread.CurrentThread.ManagedThreadId` pattern.

Reviewer: ```suggestion
<xref:System.Environment.CurrentManagedThreadId?displayProperty=fullName> is a compact and efficient replacement of the `Thread.CurrentThread.ManagedThreadId` pattern.
```

## How to fix violations

The violation can either be fixed manually, or, in some cases, using Quick Actions to fix code in Visual Studio.

The following two code snippets show a violation of the rule and how to fix it:

```csharp
using System.Threading;

class MyClass
{
    void MyMethod()
    {
        int id = Thread.CurrentThread.ManagedThreadId; // Violation occurs
    }
}
```

```vb
Imports System.Threading

Class MyClass
    Private Sub MyMethod()

>         Dim id As Integer = Thread.CurrentThread.ManagedThreadId

Reviewer: ```suggestion
        Dim id As Integer = Thread.CurrentThread.ManagedThreadId ' Violation occurs.
```

    End Function
End Class
```

```csharp
using System.Threading;

class MyClass
{
    void MyMethod()
    {
        int id = System.Environment.CurrentManagedThreadId; // Violation fixed
    }
}
```

```vb
Imports System.Threading

Class MyClass
    Private Sub MyMethod()

>         Dim id As Integer = System.Environment.CurrentManagedThreadId

Reviewer: ```suggestion
        Dim id As Integer = System.Environment.CurrentManagedThreadId ' Violation fixed.
```

    End Function
End Class
```

> [!TIP]
> A code fix is available for this rule in Visual Studio. To use it, position the cursor on the violation and press **Ctrl**+**.** (period). Choose **Use 'Environment.CurrentManagedThreadId'** from the list of options that's presented.
>
> ![Code fix for CA1840 - Use 'Environment.CurrentManagedThreadId'](media/ca1840-codefix.png)

## When to suppress warnings

> It's safe to suppress a violation of this rule if you're not concerned about the performance impact from using `Thread.CurrentThread.ManagedThreadId`

Reviewer: ```suggestion
It's safe to suppress a violation of this rule if you're not concerned about the performance impact from using `Thread.CurrentThread.ManagedThreadId`.
```
